expression/builtin_time_vec_generated.go

```diff
@@ -0,0 +1,602 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Code generated by go generate in expression/generator; DO NOT EDIT.
+
+package expression
+
+import (
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/parser/terror"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+func (b *builtinAddDatetimeAndDurationSig) vecEvalTime(input *chunk.Chunk, result *chunk.Column) error {
+	n := input.NumRows()
+
+	if err := b.args[0].VecEvalTime(b.ctx, input, result); err != nil {
+		return err
+	}
+	buf0 := result
+
+	buf1, err := b.bufAllocator.get(types.ETDuration, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf1)
+	if err := b.args[1].VecEvalDuration(b.ctx, input, buf1); err != nil {
+		return err
+	}
+
+	result.MergeNulls(buf1)
+
+	arg0s := buf0.Times()
+
+	arg1s := buf1.GoDurations()
+
+	resultSlice := result.Times()
+
+	for i := 0; i < n; i++ {
+
+		if result.IsNull(i) {
+			continue
+		}
+
+		// get arg0 & arg1
+
+		arg0 := arg0s[i]
+
+		arg1 := arg1s[i]
+
+		// calculate
+
+		output, err := arg0.Add(b.ctx.GetSessionVars().StmtCtx, types.Duration{Duration: arg1, Fsp: -1})
+		if err != nil {
+			return err
+		}
+
+		// commit result
+
+		resultSlice[i] = output
+
+	}
+	return nil
+}
+
+func (b *builtinAddDatetimeAndDurationSig) vectorized() bool {
+	return true
+}
+
+func (b *builtinAddDatetimeAndStringSig) vecEvalTime(input *chunk.Chunk, result *chunk.Column) error {
+	n := input.NumRows()
+
+	if err := b.args[0].VecEvalTime(b.ctx, input, result); err != nil {
+		return err
+	}
+	buf0 := result
+
+	buf1, err := b.bufAllocator.get(types.ETString, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf1)
+	if err := b.args[1].VecEvalString(b.ctx, input, buf1); err != nil {
+		return err
+	}
+
+	result.MergeNulls(buf1)
```

Comment:
If the computed column type is variable length, then we can't use MergeNulls. The following function is the same.